File: ruby/3.longest-substring-without-repeating-characters.rb
# @param {String} s
# @return {Integer}
def length_of_longest_substring(s)
  max = 0
  appear = {}
  window = []

  arr = s.chars
  left = 0
  right = 0

  arr.each_with_index do |char, right|
    while appear.has_key?(char)
      appear.delete(arr[left])
      left += 1
    end
    appear[char] = right
    max = [max, right - left + 1].max
  end

  max
end

File: rust/src/3.longest-substring-without-repeating-characters.rs
impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut max = 0;
        let mut appear = std::collections::HashSet::new();
        let mut window: Vec<char> = vec![];

        for char in s.chars() {
            while appear.contains(&char) {
                if let Some(c) = window.get(0) {
                    appear.remove(c);
                    window.remove(0);
                }
            }
            appear.insert(char);
            window.push(char);

            if window.len() > max {
                max = window.len();
            }
        }
        max as i32
    }
}
